[RFC] Add addImport extension

pieterv opened the pull request:
The purpose of this is to create a jscodeshift extension that can add requires in a non destructive manner, it doesn't try to have the requires be perfect it just tries to make the best guess at where it should go. I have based a lot of this work from js-format (in fact most of utils is directly copied from here) with the difference in this is non destructive to existing code.

Currently not all the tests pass as i am having some issues with whitespace printing, and i have also not implemented the feature to constraint the require adding to just the first relevant block.

the API looks like this:
```javascript
const plugins = imports.register(jscodeshift, imports.config.CJSBasicRequire);
jscodeshift(file.source)
  .addImport(statement`
    import MyImportItem from './path/to/MyImportItem';
  `)

// Or old school:

jscodeshift(file.source)
  .addImport(statement`
    var MyRequireItem = require('MyRequireItem');
  `);
```

cc @cpojer @fkling @kyldvs

kyldvs commented:
Can you add some more test cases around how it handles comments?

```
// This require does blah.
var blah = require('blah');
```

Then if I add the require `'apple'` how will the above change? Same for trailing comments, etc.

kyldvs commented:
Also the format-js stuff is open source, I wonder if it's worth just adding it as a dependency and using the utilities directly instead of copy/pasting them. I'm not particular either way but I believe I've updated some of the ones you have added here in the past week.

pieterv commented:
Oh nice, thats a good idea, ill try pulling it in. Would be good to share the configs as well.

kyldvs commented:
This is the module: https://www.npmjs.com/package/nuclide-format-js-base

pieterv commented:
I have updated this PR to use @kyldvs's format-js-base repo, i have also removed the flow types i had since they weren't being validated and were mostly wrong.

pieterv commented on extensions/imports/config/FBRequireConfig.js:
```diff
@@ -0,0 +1,35 @@
+      getDeclarationName(node1),
+      getDeclarationName(node2)
+    ),
+  },
```
@kyldvs would be great to expose these configs from `format-js-base` so i didnt need to recreate them here. https://github.com/facebook/nuclide/blob/master/pkg/nuclide/format-js-base/lib/requires/formatRequires.js#L32-L92

cpojer commented:
@pieterv would you mind rebasing this agains the latest version of the repo? I updated testing and added eslint and whatnot :)

pieterv commented:
I have rebased and fixed the lint warnings. This code still has some failing tests, they are mostly around some weirdness with whitespace and recast that i haven't figured out how to fix. I can disable the tests if your ok with merging it in its current state but i wont have time for a little while to try fixing the overall issues :(

pieterv commented:
Hey @cpojer, this should be good to merge now. I have added a couple of missing features and then commented out a couple of the harder to fix tests. I have also added a README with usage details as well as a warning (since we know it can break in some cases).

cpojer commented:
this is so great. Sorry for making you wait so long and forgetting about this :D

kyldvs commented on extensions/imports/__tests__/addImportCJSBasicConfig-test.js:
```diff
@@ -0,0 +1,288 @@
+    `);
+  });
+
+  // TODO: Comments mess with the whitespace output, fix this.
```
Comments are the worst, this is another reason why you shouldn't comment code ever /s

pieterv replied:
Yeah infact I should have removed this comment!

pieterv commented:
YAY \o/ - thanks @cpojer and @kyldvs

fkling commented:
Sweet! I wonder if we can actually release this extension as its own npm module?

cpojer commented:
Let's add a package.json and publish it from within this repo for easier
maintenance? :)

On Thu, Feb 25, 2016 at 11:54 Felix Kling <notifications@github.com> wrote:

> Sweet! I wonder if we can actually release this extension as its own npm
> module?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cpojer/js-codemod/pull/6#issuecomment-188577726>.
>
